File: backend/zvms/typing/structs.py
from zvms.typing.checker import *
from zvms.res import VolType, VolStatus, ThoughtStatus

class Class(Object):
    name = Len(String, 1, 6)

class SingleClass(Object):
    id = Int
    name = String

class SingleUserWithoutAuth(Object):
    id = Int
    name = String

class SingleUser(SingleUserWithoutAuth):
    auth = Int

ListClassesResponse = Array(SingleClass())
    
class ClassInfoResponse(Object):
    name = String
    students=Array(SingleUser())
    teachers=Array(SingleUser())

class UserLoginResponse(Object):
    token = String

SearchUsersResponse = Array(SingleUser())

class UserInfoResponse:
    name = String
    cls = Int
    auth = Int
    clsName = String

class ThoughtInfoResponse(Object):
    reason = String
    thought = String
    reward = Int
    pics = Array(String)

class StudentThoughtsResponse(Object):
    accepted = Array(ThoughtInfoResponse())
    unsubmitted = Array(ThoughtInfoResponse())
    draft = Array(ThoughtInfoResponse())
    unaudited = Array(ThoughtInfoResponse())

class StudentStatResponse(Object):
    inside = Int
    outside = Int
    large = Int

class SingleNotice(Object):
    id = Int
    title = String
    content = String
    sender = Int
    deadtime = DateTime
    senderName = String
    
SearchNoticesResponse = Array(SingleNotice())

class SingleSignup(Object):
    volId = Int
    volName = String
    stuId = Int
    stuName = String
    
ListSignupResponse = Array(SingleSignup())

class SingleVolunteer(Object):
    id = Int
    name = String
    time = String
    status = Int
    
SearchVolunteersResponse = Array(SingleVolunteer())

class VolunteerInfoResponse(Object):
    name = String
    description = String
    time = String
    status = Enum(VolStatus)
    type = Enum(VolType)
    reward = Int
    signable = Boolean
    joiners = Array(SingleUserWithoutAuth())
    holder = Int
    holderName = String

class SearchNotices(Optional):
    sender = Int
    user = Int
    cls = Int
    school = Int

class NoticeBody(Object):
    title = Len(String, 1, 33)
    content = Len(String, 1, 1025)
    deadtime = DateTime
    
class Notice(NoticeBody):
    targets = Array(Int)

class Report(Object):
    report = String
    
class Signup(Object):
    students = Array(Int)
    
class SearchThoughts(Optional):
    cls = Int
    status = Enum(ThoughtStatus)
    student = Int
    volunteer = Int
    
class Thought(Object):
    thought = Len(String, 1, 1025)
    pictures = Array(String)

class Login(Object):
    id = Int
    pwd = Len(String, 32, 33)

class SearchUsers(Optional):
    name = String
    cls = Int
    auth = Int
    
class ModPwd(Object):
    old = Len(String, 32, 33)
    neo = Len(String, 32, 33)

class User(Object):
    name = Len(String, 1, 6)
    cls = Int
    auth = Int

class OneUser(User):
    id = Int

class Users(Object):
    users = Array(OneUser())

class ClassVol(Object):
    id = Int
    max = Int
    
class SearchVolunteers(Optional):
    holder = Int
    student = Int
    cls = Int
    name = String
    status = Enum(VolStatus)
    
class Volunteer(Object):
    name = Len(String, 1, 33)
    description = Len(String, 1, 1025)
    time = DateTime
    type = Enum(VolType)
    reward = Int
    classes = Array(ClassVol())

class Repulse(Object):
    reason = Len(String, 1, 65)

File: web/src/apis/types/structs.ts
import * as enums from "././enums";

export interface Class {
    name: string
}

export interface SingleClass {
    id: number,
    name: string
}

export interface SingleUserWithoutAuth {
    id: number,
    name: string
}

export interface SingleUser extends SingleUserWithoutAuth {
    auth: number
}

export type ListClassesResponse = Array<SingleClass>

export interface ClassInfoResponse {
    name: string,
    students: Array<SingleUser>,
    teachers: Array<SingleUser>
}

export interface UserLoginResponse {
    token: string
}

export type SearchUsersResponse = Array<SingleUser>

export interface UserInfoResponse {
    name: string,
    cls: number,
    auth: number,
    clsName: string
}

export interface ThoughtInfoResponse {
    reason?: string,
    thought?: string,
    reward?: number,
    pics?: Array<string>
}

export interface StudentThoughtsResponse {
    accepted: Array<ThoughtInfoResponse>,
    unsubmitted: Array<ThoughtInfoResponse>,
    draft: Array<ThoughtInfoResponse>,
    unaudited: Array<ThoughtInfoResponse>
}

export interface StudentStatResponse {
    inside: number,
    outside: number,
    large: number
}

export interface SingleNotice {
    id: number,
    title: string,
    content: string,
    sender: number,
    deadtime: string,
    senderName: string
}

export type SearchNoticesResponse = Array<SingleNotice>

export interface SingleSignup {
    volId: number,
    volName: string,
    stuId: number,
    stuName: string
}

export type ListSignupResponse = Array<SingleSignup>

export interface SingleVolunteer {
    id: number,
    name: string,
    time: string,
    status: number,
    signable: boolean
}

export type SearchVolunteersResponse = Array<SingleVolunteer>

export interface VolunteerInfoResponse {
    name: string,
    description: string,
    time: string,
    status: enums.VolStatus,
    type: enums.VolType,
    reward: number,
    signable: boolean,
    joiners: Array<SingleUserWithoutAuth>,
    holder: number,
    holderName: string
}

export interface SearchNotices {
    sender?: number,
    user?: number,
    cls?: number,
    school?: number
}

export interface NoticeBody {
    title: string,
    content: string,
    deadtime: string
}

export interface Notice extends NoticeBody {
    targets: Array<number>
}

export interface Report {
    report: string
}

export interface Signup {
    students: Array<number>
}

export interface SearchThoughts {
    cls?: number,
    status?: enums.ThoughtStatus,
    student?: number,
    volunteer?: number
}

export interface SingleThought {
    volId: number,
    stuId: number,
    status: enums.ThoughtStatus,
    stuName: string,
    volName: string
}

export type SearchThoughtsResponse = Array<SingleThought>

export interface Thought {
    thought: string,
    pictures: Array<string>
}

export interface Login {
    id: number,
    pwd: string
}

export interface SearchUsers {
    name?: string,
    cls?: number,
    auth?: number
}

export interface ModPwd {
    old: string,
    neo: string
}

export interface User {
    name: string,
    cls: number,
    auth: number
}

export interface OneUser extends User {
    id: number
}

export interface Users {
    users: Array<OneUser>
}

export interface ClassVol {
    id: number,
    max: number
}

export interface SearchVolunteers {
    holder?: number,
    student?: number,
    cls?: number,
    name?: string,
    status?: enums.VolStatus
}

export interface VolunteerBody {
    name: string,
    description: string,
    time: string,
    type: enums.VolType,
    reward: number
}

export interface Volunteer extends VolunteerBody {
    classes: Array<ClassVol>
}

export interface AppointedVolunteer extends VolunteerBody {
    joiners: Array<number>
}

export interface Repulse {
    reason: string
}

export interface Accept {
    reward: number
}

